### PR TITLE
connector/ldap: expand LDAP connector to include searches

### PR DESCRIPTION
**Pull request by ericchiang**
wip will expand docs tomorrow and add an groups search mechanism.

cc @rithujohn191 @sym3tri

### CODE REVIEW

**Comment by ericchiang** [comment]
cc @dghubble @squat @sym3tri can you take a look at this and see if it makes sense?

I need to finish up some of the details and don't expect direct comment on the LDAP logic, but would appreciate comments on the documentation and configuration.

File: connector/ldap/ldap.go [hunk 1]
```diff
@@ -33,27 +122,185 @@ type ldapConnector struct {
...
+		// TODO(ericchiang): Is this going to spam the logs?
+		log.Printf("ldap: groups search with filter %q returned no groups", filter)
+	}
+
+	var groupNames []string
+
+	for _, group := range groups {
+		name := getAttr(*group, c.GroupSearch.NameAttr)
+		if name == "" {
+			// Be obnoxious about missing missing attributes. If the group entry is
```
Comment (ericchiang) [inline connector/ldap/ldap.go]: this seem reasonable to everyone?

File: connector/ldap/ldap.go [hunk 2]
```diff
@@ -2,20 +2,109 @@
...
 
-// Config holds the configuration parameters for the LDAP connector.
+// Config holds the configuration parameters for the LDAP connector. The LDAP
+// connectors require executing two queries, the first to find the user based on
+// the username and password given to the connector. The second to use the user
+// entry to search for groups.
+//
+// An example config:
+//
+//     type: ldap
```
Comment (ericchiang) [inline connector/ldap/ldap.go]: yay! giant configs!

**Comment by ericchiang** [comment]
FYI some background on the existing LDAP connector

* LDAP groups search is insufficient: https://github.com/coreos/dex/issues/536
* Mapping entity names to claims: https://github.com/coreos/dex/issues/552

File: connector/ldap/ldap.go [hunk 3]
```diff
@@ -2,20 +2,109 @@
...
+//     config:
+//       host: ldap.example.com:636
+//       rootCA: /etc/dex/ldap.ca
+//       bindDN: uid=seviceaccount,cn=users,dc=example,dc=com
+//       bindPW: password
+//       userSearch:
+//         # Would translate to the query "(&(objectClass=person)&(uid=<username>))"
+//         baseDN: cn=users,dc=example,dc=com
+//         filter: "(objectClass=person)"
+//         username: uid
```
Comment (dghubble) [inline connector/ldap/ldap.go]: I'm not sure how much of this is just ldap or how much is your adaptation. I think of username and uid as different

Comment (ericchiang) [inline connector/ldap/ldap.go]: FreeIPA actually uses "uid" as the human readable name you enter when doing binds. For example here's my LDAP entry:

```
dn: uid=echiang,cn=users,cn=accounts,dc=core,dc=example,dc=so
displayName: Eric Chiang
uid: echiang
objectClass: ipaobject
objectClass: person
objectClass: top
objectClass: ipasshuser
objectClass: inetorgperson
objectClass: organizationalperson
objectClass: krbticketpolicyaux
objectClass: krbprincipalaux
objectClass: inetuser
objectClass: posixaccount
objectClass: ipaSshGroupOfPubKeys
objectClass: mepOriginEntry
loginShell: /bin/bash
initials: EC
gecos: Eric Chiang
sn: Chiang
homeDirectory: /home/echiang
givenName: Eric
cn: Eric Chiang
uidNumber: 300015
gidNumber: 300015
```

File: connector/ldap/ldap.go [hunk 4]
```diff
@@ -2,20 +2,109 @@
...
 type Config struct {
-	Host   string `yaml:"host"`
+	// The host and optional port of the LDAP server. If port isn't supplied, it will be
+	// guessed based on the TLS configuration. 389 or 636.
+	Host string `yaml:"host"`
+
+	// Path to a trusted root certificate file.
+	RootCA string `yaml:"rootCA"`
+
+	// BindDN and BindPW for an application service account. The connector uses these
```
Comment (ericchiang) [inline connector/ldap/ldap.go]: Just realized that these can totally be empty. Some LDAP configurations allow enough anonymous access that you don't need a service account.

File: connector/ldap/ldap.go [hunk 5]
```diff
@@ -2,20 +2,109 @@
...
+		UserAttr  string `yaml:"userAttr"`
+		GroupAttr string `yaml:"groupAttr"`
+
+		// The attribute of the group that represents its name.
+		NameAttr string `yaml:"name"`
+	} `yaml:"groupSearch"`
 }
 
+// TODO(ericchiang): actually implement this.
+func escapeFilter(s string) string { return s }
```
Comment (ericchiang) [inline connector/ldap/ldap.go]: Still need to implement this.

File: connector/ldap/ldap.go [hunk 6]
```diff
@@ -2,20 +2,109 @@
...
+//         nameAttr: name
+//
 type Config struct {
-	Host   string `yaml:"host"`
+	// The host and optional port of the LDAP server. If port isn't supplied, it will be
+	// guessed based on the TLS configuration. 389 or 636.
+	Host string `yaml:"host"`
+
+	// Path to a trusted root certificate file.
+	RootCA string `yaml:"rootCA"`
```
Comment (ericchiang) [inline connector/ldap/ldap.go]: This is slightly different from the existing postgres args (which is `caFile`). Will try to make them more similar.

**Comment by ericchiang** [comment]
Still need to do filter escaping.

**Pull request merged**